### ISSUE TITLE
Kong 2.1.0 messes up responses when there're unhealthy upstream targets

### ISSUE THREAD

**Issue by nhamlh** [Kong/kong#6212]
### Summary
Kong 2.1.0 messes up responses when there're unhealthy upstream targets. By messing up responses I mean a client get a response of another client. When I stressed test kong 2.10 with 100 concurrent threads, then I added an unhealthy target to service's upstream. I started to observe my stress test script reporting it was receiving different response than it expected.

I also attach [a gist](https://gist.github.com/nhamlh/6898f6dee8f0c5c38bbbc6e497adcd0f) contains materials for a quick reproduce and PoC.

### Steps To Reproduce

1. Start kong 2.1.0-alpine container
2. Add a service with ring-loadbalancer
3. Stress test the service
4. Make an upstream's target unhealthy (add an invalid target or take a target down)

### Additional Details & Logs
- Kong version 2.1.0 with default configuration
- Operating system docker 19.03.12-ce

**Comment by ehdieunguyen** [comment]
@hbagdi could you take time to look this? We got this problem on production env and then we need revert Kong back to `2.0.3`. 
So, if any organizations or personal are using `2.1.0`, `2.1.1` versions, they need to know this bug soon.
Thank you so much!

**Comment by ehdieunguyen** [comment]
I also recorded a video when Kong `2.1.0` messes up responses. You can see the `request_id` of responses not match with request_id in request params; the expectation is that they need to match.

[![](http://img.youtube.com/vi/Kdt9STeCv0w/0.jpg)](http://www.youtube.com/watch?v=Kdt9STeCv0w "")

**Comment by nguyenductoan** [comment]
i can not reproduce the error with Kong `2.0.5`, so i think this error appeared from `2.1.0`

**Comment by jeremyjpj0916** [comment]
Is this when leveraging upstreams or just single backend proxies? I wonder if the new upstream pool member logic to manage connections is playing a role here otherwise. Hmm.

**Comment by nhamlh** [comment]
@jeremyjpj0916 I faced this issue with LB so I only reproduced it with LB.

**Comment by murillopaula** [comment]
If I mark a target as `unhealthy` has no effect, i.e, it's still serving requests... is this expected?

**Comment by murillopaula** [comment]
Also after I take my server down, I tried to take it back up and it seems like the round-robin LB is not actually re-proxying requests to it. Even if I mark it as `healthy` manually via the Admin API I’m not able to see requests being proxied to it.

**Comment by murillopaula** [comment]
Can this also be related? https://github.com/Kong/kong/issues/6112

**Comment by dndx** [comment]
Hey folks,

Thanks for the bug report. We are currently investigating.

**Comment by thanhnl** [comment]
I tried to reproduce this problem using Go server and client

Send this request
```
GET http://localhost:8000/?param=user-1-param-2-01efj2wetyz3b9yf7f6g0jz1j0
Accept: application/json
Header: user-1-header-2-01efj2wetyz3b9yf7f6g0jz1j0

{"body":"user-1-body-2-01efj2wetyz3b9yf7f6g0jz1j0"}
```

Server should response with
```
Header: user-1-header-2-01efj2wetyz3b9yf7f6g0jz1j0

{
  "body":"user-1-body-2-01efj2wetyz3b9yf7f6g0jz1j0",
  "header":"user-1-header-2-01efj2wetyz3b9yf7f6g0jz1j0",
  "param":"user-1-param-2-01efj2wetyz3b9yf7f6g0jz1j0",
}
```

With Kong 2.1.1, Kong did mess up the request to upstream:

```
2020/08/13 05:12:14 Unmarshal: invalid character 'G' looking for beginning of value
2020/08/13 05:12:14 body: GET /?param=user-1-param-1-01efk3qf8ta87gvew3g3p619
2020/08/13 05:12:14 header:  map[Accept-Encoding:[gzip] Connection:[keep-alive] Content-Length:[51] Header:[user-1-header-1-01efk3qf8ta87gvew3g3p6190y] User-Agent:[Go-http-client/1.1] X-Forwarded-For:[172.17.0.1] X-Forwarded-Host:[10.0.0.3] X-Forwarded-Port:[80] X-Forwarded-Prefix:[/] X-Forwarded-Proto:[http] X-Real-Ip:[172.17.0.1]]
2020/08/13 05:12:14 url:  /?param=user-1-param-1-01efk3qf8ta87gvew3g3p6190y
```

Broken body of the request that sent to upstream: `GET /?param=user-1-param-1-01efk3qf8ta87gvew3g3p619`

This problem doesn't exist in Kong 2.0.3. 

However as the go server could handle this error and response to correct request, there is no mess up between client sessions.
Here are the test code https://gist.github.com/thanhnl/3c762eb51e79cc4d8a1be691873e6b9a

**Comment by thanhnl** [comment]
After change the request to this form `GET http://localhost:8000/user-1-param-2-01efj2wetyz3b9yf7f6g0jz1j0` , the problem come clearly in Kong 2.1.1
```
## mismatch header:                                                                                                                                                                                                                         
request : user-1-header-0-01efkwt9wsbpxnc03dcdfx7td3                                                                                                                                                                                        
response: map[Content-Length:[117] Content-Type:[application/json] Date:[Thu, 13 Aug 2020 12:30:41 GMT] Header:[user-3-header-9-01efkwt9kh2jrnt5q8v9n4xmp1] Via:[kong/2.1.1] X-Kong-Proxy-Latency:[3] X-Kong-Upstream-Latency:[194]]
request:  {user-1-body-0-01efkwt9wsbpxnc03dcdfx7td3 user-1-param-0-01efkwt9wsbpxnc03dcdfx7td3 user-1-header-0-01efkwt9wsbpxnc03dcdfx7td3}                                                                                                   
body   :  {"body":"","param":"user-3-param-9-01efkwt9kh2jrnt5q8v9n4xmp1","header":"user-3-header-9-01efkwt9kh2jrnt5q8v9n4xmp1"}        
                                                                                                     
## mismatch param:                                                                                                                                                                                                                          
request : user-1-param-0-01efkwt9wsbpxnc03dcdfx7td3                                                                                                                                                                                         
response: user-3-param-9-01efkwt9kh2jrnt5q8v9n4xmp1
request:  {user-1-body-0-01efkwt9wsbpxnc03dcdfx7td3 user-1-param-0-01efkwt9wsbpxnc03dcdfx7td3 user-1-header-0-01efkwt9wsbpxnc03dcdfx7td3}                                                                                                   
header :  map[Content-Length:[117] Content-Type:[application/json] Date:[Thu, 13 Aug 2020 12:30:41 GMT] Header:[user-3-header-9-01efkwt9kh2jrnt5q8v9n4xmp1] Via:[kong/2.1.1] X-Kong-Proxy-Latency:[3] X-Kong-Upstream-Latency:[194]]        
body   :  {"body":"","param":"user-3-param-9-01efkwt9kh2jrnt5q8v9n4xmp1","header":"user-3-header-9-01efkwt9kh2jrnt5q8v9n4xmp1"} 
```
 Test code https://gist.github.com/thanhnl/03033ab9911d168d97955e8f9f532270

**Comment by dndx** [comment]
Hello everyone,

Thanks again for the detailed bug report and the reproduction steps. We can confirm the bug indeed occurs in Kong version > 2.1.0.

This bug could be mitigated temporarily by disabling upstream keepalive pools. It can be achieved by either:
1. In `kong.conf`, set `upstream_keepalive_pool_size=0`, or
2. Setting the environment `KONG_UPSTREAM_KEEPALIVE_POOL_SIZE=0` when starting Kong with the CLI.

Please know that we are taking this bug very seriously. A new release containing the fix will be out as soon as possible. Please stay tuned.

**Comment by nhamlh** [comment]
After a more comprehensive test as follow:
Clients send a bunch of request to a service behind kong with a request param, a header, and a body with the same value. Upstream service does an integrity check to these value, it'll output a debug log if the check fails. Upstream service then construct a json object of those values and return to clients. Clients also perform an integrity check with the response.

It looks like we're facing two different bugs:
1. When clients send requests **without body** while there're unhealthy upstream targets. Clients start to get response of others.
Each request contains a tuple of param and header with the same value. The integrity check of server all passes.

2. When clients send requests ** body** while there're unhealthy upstream targets. Clients get back its own request, but the response body is a mess. The integrity check of server fails. More specifically, kong messed up the request body before sending it to upstream servers.

I updated the gist to quickly reproduce the cases.

**Comment by nhamlh** [comment]
Oh thank @dndx

**Comment by nhamlh** [comment]
@dndx I just want to ask again to confirm if the bug occurs in kong = 2.1.0 as you said kong > 2.1.0

**Comment by dndx** [comment]
@nhamlh Yes, it occurs in Kong >= 2.1.0. Sorry for the confusion, I have updated the comment.

**Comment by dndx** [comment]
@nhamlh If you are able, could you please give a try of https://github.com/Kong/kong/pull/6224? It should have fixed this issue.

**Comment by dndx** [comment]
Hello everyone,

Thanks for your patience! We have released Kong 2.1.2 patch version which should fix this issue. Please give it a try.

Changelogs for 2.1.2: https://github.com/Kong/kong/blob/master/CHANGELOG.md#212

**Comment by nhamlh** [comment]
Thank @dndx! I've just tested again with kong 2.1.2 and it looks like the bug is gone. I'm closing this issue now.

**Comment by lclin412** [comment]
> Hello everyone,
> 
> Thanks for your patience! We have released Kong 2.1.2 patch version which should fix this issue. Please give it a try.
> 
> Changelogs for 2.1.2: https://github.com/Kong/kong/blob/master/CHANGELOG.md#212



> Thank @dndx! I've just tested again with kong 2.1.2 and it looks like the bug is gone. I'm closing this issue now.


I still encounter this problem in version 2.2.0. Is this problem solved in this version (2.2.0)? Thanks for the help.
